Backports for 0.3X_experimental_ai5

andrewkaufman opened the pull request:
I went through all the 0.4X releases notes and cherry-picked everything that seemed production relevant and didn't have big conflicts. I think I do have a couple breaking changes in here, but they're hopefully ones without a big impact.

@johnhaddon can you have a look at what I've grabbed and see if anything stand out as an issue?

The things I wanted to cherry-pick but bailed out of are:
- TaskNode : Errors are now reported in the NodeGraph (#2163)
- Shader/ShaderPlug : Fixed GIL management bug which could cause a deadlock
in the attributes() and attributesHash() methods (#2354).
- SceneInspector (#2222)

andrewkaufman commented:
Also note that I still need to do some interactive testing to see if the fixes actually work on this branch.

johnhaddon commented on include/Gaffer/ScriptNode.h:
```diff
@@ -182,7 +182,7 @@ class ScriptNode : public Node
 		bool isExecuting() const;
 		/// This signal is emitted following successful execution of a script.
-		typedef boost::signal<void ( ScriptNodePtr, const std::string )> ScriptExecutedSignal;
+		typedef boost::signal<void ( ScriptNode *, const std::string )> ScriptExecutedSignal;
```
Unless we have a very good reason, we should omit this commit because it's an ABI change.

andrewkaufman replied:
Can you remind me what are the symptoms of the bug this was fixing? If needed I can make it a 0.36 release. Maybe it was a bug that wasn’t affecting IE but an external user?

johnhaddon replied:
This commit was included in #2336, which fixed an expression copy/paste bug. I don't think the commit is part of the bug fix though, it's just something I fixed up along the way. From that PR :

>I've also included a fix for the ScriptNode::ScriptExecutedSignal signature because it tripped me up while writing this. This could be considered separately if you prefer.

When I said "tripped me up" I think I just meant "the signature was not what I expected it to be" rather than anything more important.

andrewkaufman replied:
I dropped the commit and rebased

andrewkaufman commented:
Also, did I manage to grab the fix for those image node serialization issues, or is that in the ValuePlug PR I bailed out from? If so, maybe I need to make another stab at getting that working.

johnhaddon commented:
> Also, did I manage to grab the fix for those image node serialization issues, or is that in the ValuePlug PR I bailed out from?

Yeah, I think that's #2395. That has an ABI change in it too, because it removes a virtual method from ValuePlugSerialiser. You should get away with that one though because it doesn't look like we're using that class in any IE code. Since this isn't an official release branch I think it'd be OK to take a liberty there if it makes deployment easier, but for any other branch we should play it by the book.

andrewkaufman commented on python/GafferTest/ExpressionTest.py:
```diff
@@ -1329,5 +1329,22 @@ def testCopyPaste( self ) :
 			self.assertEqual( s["n"]["sum"].getValue(), 32 )
 			self.assertEqual( s["n1"]["sum"].getValue(), 35 )
 
+	def testNonSerialisableInput( self ) :
```
I tweaked the test in e1dffcd

andrewkaufman replied:
Actually, scratch that. dbf62d0 changes those values anyways, presumably to employ this same workaround.

andrewkaufman commented:
I had to fixup some more things for that ValuePlug commit. I think its working now, but I'm a little uncertain if I've done the right things here: d383ed27584369d38dcce61e7958c9a66a982307

andrewkaufman commented:
Also note that I've kept `static std::string repr( const ValuePlug *plug )` though it doesn't seem to be used anywhere, nor does it exist in master currently...

johnhaddon commented:
> I had to fixup some more things for that ValuePlug commit. I think its working now, but I'm a little uncertain if I've done the right things here: d383ed2

That looks right to me.

> Also note that I've kept static std::string repr( const ValuePlug *plug ) though it doesn't seem to be used anywhere, nor does it exist in master currently...

That's referenced by the `.def( "__repr__", &repr )` line isn't it?

andrewkaufman commented:
Oh yeah, you're right. And on master it seems that's all happening in another file now: https://github.com/GafferHQ/gaffer/blob/master/src/GafferModule/ValuePlugBinding.cpp

johnhaddon commented:
> Oh yeah, you're right. And on master it seems that's all happening in another file now

Yep. GafferBindings now contains only the stuff that needs to be public - just the bits needed to bind your own types. GafferModule contains all the internal machinery necessary for the Gaffer python module, but which doesn't need to be exposed anywhere...

andrewkaufman commented:
Hmm, failing in the GafferRenderMan tests now, which I'm unfortunately not able to run at IE right now. Any ideas?

```
======================================================================
ERROR: testConvertFixedCoshaderArrayToDynamicDuringLoading (GafferRenderManTest.RenderManShaderTest.RenderManShaderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/GafferHQ/gaffer/install/gaffer-0.35.1.2-linux/python/GafferRenderManTest/RenderManShaderTest.py", line 1257, in testConvertFixedCoshaderArrayToDynamicDuringLoading
    s2.execute( s.serialise() )
RuntimeError: Exception : Line 27 : RuntimeError: Exception : Plug "ScriptNode.n.parameters.fixedShaderArray.fixedShaderArray1" rejects input "ScriptNode.c.out".


======================================================================
ERROR: testSerialiseDynamicCoshaderArrayParameters (GafferRenderManTest.RenderManShaderTest.RenderManShaderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/GafferHQ/gaffer/install/gaffer-0.35.1.2-linux/python/GafferRenderManTest/RenderManShaderTest.py", line 1166, in testSerialiseDynamicCoshaderArrayParameters
    s2.execute( s.serialise() )
RuntimeError: Exception : Line 30 : RuntimeError: Exception : Plug "ScriptNode.n.parameters.dynamicShaderArray.dynamicShaderArray0" rejects input "ScriptNode.c.out".


----------------------------------------------------------------------
```

andrewkaufman commented:
Also note I've added 3 extra commits here (not todo with those test failures).

andrewkaufman commented:
I force rebased having:

- dropped a SceneInspector SetHistory commit
- dropped the contentious ValuePlug serialisation commits
- added BoxIO value promotion (from  #2461)
- added Tractor plug serialisation (from #2463)

andrewkaufman commented:
Mac travis tests are failing with some tslv1 issue, but this is the backport-only-for-IE branch, so I'm going to ignore it.

I've also cherry-picked John's script frame stuff and Dan's graph component name clash stuff.
